pkg/datapath/ipcache/listener.go

```diff
@@ -300,24 +306,35 @@ func (l *BPFListener) garbageCollect(ctx context.Context) (*sync.WaitGroup, erro
 		// will pick up the new paths without requiring recompilation.
 		backupMapName := fmt.Sprintf("%s_old", ipcacheMap.Name)
 		if err := shuffleMaps(ipcacheMap.Name, backupMapName, pendingMapName); err != nil {
+			ipcache.IPIdentityCache.RUnlock()
 			return nil, err
 		}
 
-		wg, err := l.datapath.TriggerReloadWithoutCompile("datapath ipcache")
-		if err != nil {
+		// Reopen the ipcache map so that new writes and reads will use
+		// the new map
+		if err := ipcacheMap.Reopen(); err != nil {
 			handleMapShuffleFailure(backupMapName, ipcacheMap.Name)
+			ipcache.IPIdentityCache.RUnlock()
 			return nil, err
 		}
 
-		// If the base programs successfully compiled, then the maps
-		// should be OK so let's update all references to the IPCache
-		// so that they point to the new version.
-		_ = os.RemoveAll(bpf.MapPath(backupMapName))
-		if err := ipcacheMap.Reopen(); err != nil {
-			// Very unlikely; base program compilation succeeded.
-			log.WithError(err).Warning("Failed to reopen BPF ipcache map")
-			return nil, err
+		// Unlock the ipcache as in order for
+		// TriggerReloadWithoutCompile() to succeed, other endpoint
+		// regenerations which are blocking on the ipcache lock may
+		// need to succeed first (#11946)
+		ipcache.IPIdentityCache.RUnlock()
+
+		wg, err := l.datapath.TriggerReloadWithoutCompile("datapath ipcache")
+		if err != nil {
+			// We can't really undo the map rename again as ipcache
+			// operations had already been permitted so the backup
+			// map is potentially outdated. Fail hard to restart
+			// the agent so we reconstruct the ipcache from
+			// scratch.
+			log.WithError(err).Fatal("Endpoint datapath reload triggered by ipcache GC failed. Inconsistent state.")
 		}
+
+		_ = os.RemoveAll(bpf.MapPath(backupMapName))
```

Comment:
Should we do this before the compile (or at least in the error case) to ensure that in the unlikely event that we `Fatal()` out above, the old map is removed from the filesystem?

Reply:
I guess that assuming the new run of the daemon executes through this path successfully, it will remove the old map 5 minutes after startup.

Reply:
Given the `Rename()` behaviour described below + the periodic deletion / recreate, I don't think this is blocking.

https://github.com/cilium/cilium/pull/11950#discussion_r436813198

Reply:
I've discussed this with Jarno as well. I've left it there on purpose that in the event that the Fatal occurs, the map is left behind so we can potentially troubleshoot it.

The next GC run will remove it as the agent restarts so it will only persist if the system remains in an inconsistent state.

Reply:
I asked the same question. Turns our that `os.Rename()` will replace the old file if it exists, and there is value of having the old file left behind in case you get to debug the error case.